### ISSUE TITLE
orderbook sell - odd behavior

### ISSUE THREAD

**Issue by xmatthias** [freqtrade/freqtrade#1667]
##  Describe your environment

  * Python Version: 3.7.0
  * CCXT version: 1.18.367 (`pip freeze | grep ccxt`)
  * Branch: Develop
  * Last Commit ID: d66e6510e30c65cd8159104b688ae433b98927c5 
  
## Step 3: Describe the problem:

Using orderbook with the following ask strategy, the bot buys and sells almost immediately with ROI as sell reason.

Configuration:
```
    minimal_roi = {
        "0": 0.01
    }
```

``` json
"ask_strategy":{
    "use_order_book": false,
    "order_book_min": 1,
    "order_book_max": 10
}
```

The relevant code seems to be the following:
https://github.com/freqtrade/freqtrade/blob/71c530590eb4080d901c11075ff577e9dc6296c9/freqtrade/freqtradebot.py#L600-L614 

It walks the ask-side and once it finds a spot matching ROI / sell-signal places the sell there.

In production mode, the order will either be fullfilled once it reaches this limit - or time out (more likely - probably).

In dry/validation mode however it'll produce very odd and misleading results since it assumes that rate reached - if i set my ROI to 10% - and walk 100 steps up the orderbook i may find that filled immediately as well - depending on the pair.

![2019-03-16-131755_537x228_scrot](https://user-images.githubusercontent.com/5024695/54475120-f8e3a300-47ed-11e9-9eb3-3b94fb957987.png)

Notice the difference between limit and "current rate" (profit uses limit for the calculation ...).

### Relevant code exceptions or logs:

  ```
2019-03-16 13:14:50,378 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 0.1 ...
2019-03-16 13:14:50,378 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2019-03-16 13:14:50,692 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'AST/ETH', 'limit': 0.0002841, 'stake_amount': 0.1, 'stake_currency': 'ETH', 'fiat_currency': 'USD'}
2019-03-16 13:14:50,816 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:50,829 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:52,678 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now)
2019-03-16 13:14:52,679 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now) failed: myTrade-Dict empty found
2019-03-16 13:14:52,680 - freqtrade.persistence - INFO - Updating trade (id=1) ...
2019-03-16 13:14:52,680 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now).
2019-03-16 13:14:52,712 - freqtrade.freqtradebot - INFO - Using order book for selling...
2019-03-16 13:14:53,088 - freqtrade.freqtradebot - INFO -   order book asks top 1: 0.00028410
2019-03-16 13:14:53,089 - freqtrade.freqtradebot - INFO -   order book asks top 2: 0.00028440
2019-03-16 13:14:53,089 - freqtrade.freqtradebot - INFO -   order book asks top 3: 0.00028540
2019-03-16 13:14:53,090 - freqtrade.freqtradebot - INFO -   order book asks top 4: 0.00028570
2019-03-16 13:14:53,090 - freqtrade.freqtradebot - INFO -   order book asks top 5: 0.00028630
2019-03-16 13:14:53,091 - freqtrade.freqtradebot - INFO -   order book asks top 6: 0.00028770
2019-03-16 13:14:53,098 - freqtrade.freqtradebot - INFO - Using order book for selling...
2019-03-16 13:14:53,526 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': sell, 'exchange': 'Binance', 'pair': 'AST/ETH', 'gain': 'profit', 'limit': 0.0002877, 'amount': 351.98873636043646, 'open_rate': 0.0002841, 'current_rate': 0.0002841, 'profit_amount': 0.00106589, 'profit_percent': 0.01064827, 'sell_reason': 'roi', 'stake_currency': 'ETH', 'fiat_currency': 'USD'}
2019-03-16 13:14:53,617 - freqtrade.freqtradebot - INFO - executed sell, reason: SellType.ROI
2019-03-16 13:14:53,618 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:53,624 - freqtrade.wallets - INFO - Wallets synced ...

```

**Comment by mishaker** [comment]
Unless I am missing something but it is plain wrong: 
when you wanna sell you don't look at asks, you look at bids ... 
you go "down" the bids and try to match your sell volume with sum of the N first bids to maximize the possibility of your order being executed. 
This code is doing quite the opposite:
For selling the position it looks for a seller ! (asks side). and it goes "up" the ladder, so it would most probably find a seller willing to sell at a price matched with TA and then tries to sell to that seller :))) 
So basically it places a take profit order hoping that the seller will be converted to a buyer with the same price over the timeout period :))))

**Comment by xmatthias** [comment]
it's not that wrong for sell eventually ....  I think it does make (kind of) sense to look at asks - if you assume the price will go up and you look for a "spot to sell" / place your limit-sell-order early. 

In the context of the bot, it'll act strangely, since it'll send the notification before the price is actually reached - but from a trading perspective, it can make sense.

Looking at more than the first "bid" while trying to sell does not make much sense - 
assume you place a sell order on the 3rd bid-spot - you've just given away profit of 2 slots, which would have bought from you...

Unfortunately, we lack clear documentation on the orderbook feature, which should clarify/explain if this is intended and why.

However - I am looking for inputs on this, hence this issue ...

**Comment by mishaker** [comment]
>it's not that wrong for sell eventually .... I think it does make (kind of) sense to look at asks - if you assume the price will go up and you look for a "spot to sell" / place your limit-sell-order early.

I don't get it. how is it different from take profit? if you find a spot at pass your order at the exact price you are behind the queue. so chances are your order is not fulfilled. 

>Looking at more than the first "bid" while trying to sell does not make much sense -
assume you place a sell order on the 3rd bid-spot - you've just given away profit of 2 slots, which would have bought from you...

If makes sense when you consider the volume. maybe the best bid is less than your volume. so you have to go down.

**Comment by xmatthias** [comment]
>  how is it different from take profit? if you find a spot at pass your order at the exact price you are behind the queue. so chances are your order is not fulfilled.

it really isn't that different...
The difference is that you're "in the middle" of the queue ... assuming you placed the order 10 slots above ... other orders will probably come in after you (and other orders have been there since 2 hours...)

> If makes sense when you consider the volume. maybe the best bid is less than your volume. so you have to go down.

trading-technically yes - but volume is not considered in the orderbook sell logic.

**Comment by mishaker** [comment]
>The difference is that you're "in the middle" of the queue ... assuming you placed the order 10 slots above ... other orders will probably come in after you (and other orders have been there since 2 hours...)

I am still not convinced, and I don't think that it was the purpose of this code. maybe we should ask @nullart2 ?

>trading-technically yes - but volume is not considered in the orderbook sell logic.

Right. what I meant was that it would have made sense if it were implemented like that, but it's not the case anyway.

**Comment by mishaker** [comment]
the order book buy strategy doesn't make sense a ton either: 

https://github.com/freqtrade/freqtrade/blob/ff08416b12aedc564a14441a426be4d807861c3f/freqtrade/freqtradebot.py#L260-L269

It uses the top bid price to place the buy order ... which should be logically last "ask" price. 

I am half tempted to remove `use_order_book` functionality as it is used today. 
What do you think ?

**Comment by xmatthias** [comment]
i'm not sure ... honestly it simulates a potentially valid manual trading behavior ... 

you look at the bid orderbook - take a spot "a few" slots lower and place your buy there (normally right above a high volume ) ... and same for sell, but with asks.

**Comment by iuvbio** [comment]
hi, I was looking at this a while back and at first I was very surprised, because I thought it should be like @mishaker says
> when you wanna sell you don't look at asks, you look at bids ...
you go "down" the bids and try to match your sell volume with sum of the N first bids to maximize the possibility of your order being executed.

However, after thinking about it, I came to the same conclusion as @xmatthias that it does make sense, but only if you believe the price to go up/down for sells/buys. In any case, I think there should at least be an option or it should be dynamically set. As it is currently, it is definitely more risky, but more profitable. If you look at bids, you're pretty much sure that your order is filled right away. Intuitively, it makes more sense to look at bids for sells, but I do see the reasoning behind using asks.. it would be good to hear from the creator if this was intentional.

**Comment by mishaker** [comment]
I am not in front of my laptop but this conversation is becoming
interesting ...
I am trying to find a reason why it is implemented like that.
The only reason I see is to adjust take profit against order timeout. that
is:
Given that my GTC sell order (historically in FT only GTCs were supported)
has a timeout of x seconds, at what price should I push my sell order to be
almost sure it is fulfilled AND the price would have met the TA condition
(each ask price is processed again in TA).
If this is true then it is at best too complicated for not a big added
value.
I can understand the reason why someone would implement that, but I don’t
see how it can be popular.
Orderbook data should be passed to the strategy via dataprovider. There is
no reason the decision should be taken in freqtradebot.py

**Comment by bparlan** [comment]
> when you wanna sell you don't look at asks, you look at bids ...
> you go "down" the bids and try to match your sell volume with sum of the N first bids to maximize the possibility of your order being executed.

I came here to find any way to replicate this behavior. still not sure whether order-book function does it, since this is a bug report topic, but  will definitely try to make it work.
